manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,244 @@
+package deallocator
+
+import (
+	"reflect"
+
+	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/store"
+	"golang.org/x/net/context"
+)
+
+var (
+	resourceDeallocators = make(map[reflect.Type]resourceDeallocator)
+)
+
+func registerServiceLevelResource(resourceDeallocator resourceDeallocator) {
+	resourceDeallocators[reflect.TypeOf(resourceDeallocator.event)] = resourceDeallocator
```

Comment:
I don't think there's any major performance overheads, but I don't know i this is a good pattern. cc @dperny

Reply:
I also tend to shy away from using reflection when I can, but I couldn't find any other way to achieve what I wanted here, namely: have different callbacks for different event types _without_ hard-coding in this file all the event types and their callback - for the sake of cleanly separating the deallocator itself from the deallocation code for each type of resource.

But then I'd love to be hinted at a better way of doing this :) !